### DISCUSSION TITLE
Create New Plugin For Autodiscovery

### DISCUSSION THREAD

**Discussion by mr1716** [netbox-community/netbox#12648] [Ideas]
Create New Plugin For Autodiscovery. This request is for a new plugin that would allow for Netbox to automatically detect new devices on the network such as WAPs, Switches, and such. This would make it easier to add devices to Netbox. Rather than having to manually add these devices in, a plugin to automatically add these devices, would be amazing!

**Comment by candlerb** [comment]
FYI, that's not how Netbox is designed to be used:
https://docs.netbox.dev/en/stable/introduction/#serve-as-a-source-of-truth

> NetBox intends to represent the *desired* state of a network versus its *operational* state. As such, automated import of live network state is strongly discouraged. All data created in NetBox should first be vetted by a human to ensure its integrity. NetBox can then be used to populate monitoring and provisioning systems with a high degree of confidence.

However, a tool which imports data into the new Remote Data Sources feature in Netbox, compares it with the main Netbox database, highlights differences, and lets you click a button to selectively import, could be a time-saver when first setting up a network.

Note that in general it's not possible to do this accurately without human intervention. For example, Netbox makes a hard distinction between a "Device" and a "Virtual Machine"; these are separate objects and you cannot convert one into the other without deleting and recreating.  But any system which auto-discovers a "thing" on the Netbox is unlikely to be able to determine for itself which one it is: the whole point of VMs is that they look very much like real machines.

**Comment by fxrsliberty** [comment]
Without an auto discovery feature, netbox is no better then a fancy spreadsheet. Especially if there is no internal confirmation of items added. The idea that it is a vision of what is wanted, doesn't make it "true". So it's not even a "source of truth" it's an "I wish this was true"

**Reply by candlerb** [reply]
Yes, that's exactly what it is: the *desired* state of the network, which can then be used to drive automation to make it so. It's not a monitoring system.
